### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by zschira**
Adds an initial set of tests to the archiver.

Note: Zenodo tests will likely change substantially during future refactor.

### CODE REVIEW

File: tests/integration/zenodo_test.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+"""Test zenodo client api."""
```
Comment (zschira) [inline tests/integration/zenodo_test.py:1]: This test occasionally raises an exception from line 284 of `api_client.py`:

`Error Zenodo should only return a single deposition`

This error does not occur all of the time however, and should be investigated further.

**Review by jdangerx — CHANGES_REQUESTED** [review]
This all looks pretty good! A few quality-of-life nits that you can choose to pick or not, and one blocking change request for stronger assertions around our file download behavior.

Kind of stinks that we have to make a new sandbox Zenodo dataset every time we run that integration test, though it's not particularly avoidable with the "once you publish, you can't delete" thing.

I guess we could just keep publishing new versions to the same dataset, and publishing a "cleaned/almost-empty" version after each test - that's kind of worse, we'd run into all sorts of race conditions when two people are running tests at the same time.

Or we could avoid the publishing altogether, but again we'd run into race conditions if two people are trying to edit the same unpublished deposition at the same time.

File: tests/integration/zenodo_test.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        resources = {
+            file_data["path"].name: ResourceInfo(
+                local_path=file_data["path"], partitions={}
+            )
+            for file_data in test_files["original"]
+        }
+        await interface.add_files(resources)
+
+    # Wait before trying to access newly created deposition
+    time.sleep(1.0)
```
Comment (jdangerx) [inline tests/integration/zenodo_test.py:177]: If we add smart retrying logic in the future we can probably avoid these `time.sleep`s.

File: tests/integration/zenodo_test.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    async with client.deposition_interface(
+        "pudl_test", initialize=True, dry_run=False
+    ) as interface:
+        resources = {
+            file_data["path"].name: ResourceInfo(
+                local_path=file_data["path"], partitions={}
+            )
+            for file_data in test_files["original"]
+        }
+        await interface.add_files(resources)
```
Comment (jdangerx) [inline tests/integration/zenodo_test.py:174]: I got some errors from Zenodo here where the bucket API was giving me "METHOD_NOT_ALLOWED" - I put a quick fix into #23 ( in [this commit](https://github.com/catalyst-cooperative/pudl-archiver/pull/23/commits/397e78b870302e098a4432569b1727ec06702eb5) ), but hopefully that will go away as we start to pull the logic apart.

File: tests/unit/archive_base_test.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+from pathlib import Path
+from zipfile import ZipFile
+
+import pytest
+
+from pudl_archiver.archivers.classes import AbstractDatasetArchiver, ArchiveAwaitable
+
+
+@pytest.fixture(name="bad_zipfile")
+def create_bad_zipfile():
```
[the remark this answers is not in the document]
Comment (zschira) [reply tests/unit/archive_base_test.py]: No reason in particular. I think not renaming unless there's a conflict probably makes more sense, so I've changed this

File: tests/unit/archive_base_test.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    """Create a fake bad zipfile in temporary directory."""
+    with tempfile.TemporaryFile() as file:
+        file.write(b"Fake non-zipfile data")
+        yield file
```
[the remark this answers is not in the document]
Comment (zschira) [reply tests/unit/archive_base_test.py]: Yeah I was thinking about adding type hints to make the inconsistency clear, but just making these both yield filenames seems better

File: tests/unit/archive_base_test.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        file = file_path.open("rb")
+        assert file.read() == file_data
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "html,pattern,links",
+    [
+        (
+            """
```
[the remark this answers is not in the document]
Comment (zschira) [reply tests/unit/archive_base_test.py]: I like this idea. I think being able to add more html docs to the test cases is important here, but this approach does seem much cleaner

File: tests/integration/zenodo_test.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+from pudl_archiver.archivers.classes import ResourceInfo
+from pudl_archiver.frictionless import DataPackage
+from pudl_archiver.zenodo.entities import DepositionCreator, DepositionMetadata
+
+
+@pytest.fixture(name="dotenv")
+def dotenv():
+    """Load dotenv to get API keys."""
+    load_dotenv()
```
[the remark this answers is not in the document]
Comment (zschira) [reply tests/integration/zenodo_test.py:22]: How exactly do you mean? Personally I have one `.env` file with production and sandbox keys on my local machine

File: tests/integration/zenodo_test.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+
+@pytest.fixture(name="deposition_metadata")
+def deposition_metadata():
+    """Create fake DepositionMetadata model."""
```
[the remark this answers is not in the document]
Comment (zschira) [reply tests/integration/zenodo_test.py:27]: Yeah I'm good with not requiring docstrings in tests, updated

File: tests/integration/zenodo_test.py (hunk 3 above)
Comment (zschira) [reply tests/integration/zenodo_test.py:174]: Hmm, again I'm not seeing that error arise for me. I wonder if there's some system configuration weirdness that's causing some of these inconsistencies?

Comment (jdangerx) [reply tests/integration/zenodo_test.py:174]: This was due to some bucket/file API weirdness - bucket API expects `PUT` and file API expects `POST`.

File: tests/integration/zenodo_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jdangerx) [reply tests/integration/zenodo_test.py:22]: I was thinking we might have this assert that there's no production keys in the environment, and have a special .env file only for testing. 

I think the value is pretty small. Better would be to use some sort of secret manager, put our secrets in there, and have [dotenv read from the network](https://github.com/theskumar/python-dotenv#parse-configuration-as-a-stream) - obviously not within the scope of this PR!

File: tests/integration/zenodo_test.py (hunk 1 above)
Comment (jdangerx) [reply tests/integration/zenodo_test.py:1]: We fixed this - the culprit was:
1. we published a deposition and, soon after, created a new version
2. When we tried to get the deposition for the concept DOI, the old published version and the new unpublished one were both returned
3. We chose to pick the newer one (based on the DOI monotonically increasing)

**Review by jdangerx — APPROVED** [review]
After extensive pairing, I think we got the flakiness out of these tests!

Using

```python
@pytest.mark.parameterize("execution_number", range(100))
def test_zenodo...
```

I was able to run the test 100 times, and got:

```
=============================================== 100 passed, 101 warnings in 3124.41s (0:52:04) ================================================
```

Hooray! And I guess now we know our test suite runs in 31.24 seconds.

**Pull request merged**